### ISSUE TITLE
New Wallet Development Direction? Improve Bitcoin.org Inclusion Scoring?

### ISSUE THREAD

**Issue by nopara73** [zkSNACKs/Meta#43]
While we seem to be eligible for Bitcoin.org inclusion, I think it makes sense to bring closer improvements we were planning to do anyway, but we were not considering them as priority. It may be the right timing, because we have just completed all our "priority" issues we had and we are looking for a new direction to take our development. And you know me guys, I love to work on the basics, so working on it can be highly beneficial in the long term.  
In this issue I gather all the possible improvements that can potentially improve our scoring in the website. I also include improvements and tasks those, while do not help to improve scoring, but the review of @crwatkins and @harding [identified](https://github.com/bitcoin-dot-org/bitcoin.org/pull/2788) as possible improvements to our wallet.

Before that, first question should be for us to decide if this goal is worth pursuing. Please see the "Alternatives" section at the end and share your opinions @lontivero @danwalmsley @molnard @hgergoil @bharmat 

### a) Level.

Current: 3.

> Level: Each wallet must have a level property assigned. A value must be in a range between 1 and 4. Level represents a category of a wallet:  
> - Level 1 - Full nodes  
> - Level 2 - SPV, Random servers  
> - Level 3 - Hybrid, Multisig wallets  

We could easily make Wasabi SPV, we could just synchronize the header chain with NBitcoin, that's how HiddenWallet worked. However I believe that would be less secure, than our current way of doing things, since then Wasabi users could be Sybil attacked or miner attacked. Now, if a node sends us a block that does not match the hash we get from the backend, then we don't accept the block. This applies to our new partial Bitcoin Core integration, thus the local node is untrusted. I've seen @NicolasDorier and @petertodd share a similar concerns with SPV nodes. Thus I think working on level 2 would not make sense.  

Thus if we want to improve the level we need a full node integration. One way to go about it could be to use Stratis's C# Bitcoin full node code: https://github.com/stratisproject/StratisBitcoinFullNode or go with `bitcoind`. While the first option is appealing, many users use and trust more `bitcoind` already, and it's probably more stable so it may be smarter to go down on that path. The ceveat is the difficulty of handling Core's constant changes. Shit can happen if the user upgrades its own node, but not its built-in Wasabi Core node, or the other way around. Also note that we need built-in Wasabi node. Maybe we could do the same way as we do with Tor, if the user has Tor we use that, if he doesn't we use Wasabi's. Anyway, it makes this complicated, but it's not unsolvable. It's just a detail.

I imagine this integration as an opt-out integration. Until Bitcoin Core is not fully synced, Wasabi would be used in light wallet mode, like how @jonasschnelli built his [full-spv Bitcoin Core PR](https://github.com/bitcoin/bitcoin/pull/9076). However the a dialog should be shown to the user at first start that "hey, Wasabi is going to synchronize Bitcoin Core, and will work in light wallet mode until it's fully synced" and then buttons: "ok" and "no i want light wallet mode only."  

### b) Validation

Currently: `checkfailvalidationcentralized`. I believe the above improvement would make it `checkneutralvalidationvariable`. Or is it `checkgoodvalidationfullnode`? Not sure. @crwatkins can you help me out with this? 

## Deterministic Build

Current: `checkpasstransparencyopensource`. We could improve it with adding deterministic builds: `checkgoodtransparencydeterministic`. I still have my doubts as to how useful it is, but at the very least we can research into this topic and even if our opinion does end up it considering it unnecessary we can still add it if it does not come with considerable administration overhead or caveats: https://github.com/zkSNACKs/WalletWasabi/issues/1078

## Two-Factor-Auth

Current: `checkfailenvironmentdesktop`. This is another thing I'm not sure about if we can improve upon it. There are two levels we can target: `checkgoodenvironmenthardware` and `checkpassenvironmenttwofactor`.  
@crwatkins can you help me out with this? Based on [Electrum's scoring](https://bitcoin.org/en/wallets/desktop/windows/electrum/) I see that opt-in two factor makes it eligible to `checkpassenvironmenttwofactor`, but opt-in hardware wallet integration does not make it eligible to `checkgoodenvironmenthardware`. Is that correct? If so, the best course of action is to add 2FA to Wasabi.  
One thing to note though. We must be very careful with 2FA, because as Electrum's developer @ecdsa noted in Building on Bitcoin "users are more often losing their passwords than getting hacked." This corresponds with our experience, we didn't even yet have any user of ours who got hacked but we already had a handful of them who lost access to the wallet. (Some succeeded to gain access though.) I don't want us to end up doing more harm than good. Maybe we should consider building in our password typo fixer/bruteforcer to improve the situation before we ruin it with 2FA: https://github.com/zkSNACKs/WalletWasabi/issues/1035  

## Network Level Privacy

I don't want to go into the details here, rather I just refer to the conversation: https://github.com/bitcoin-dot-org/bitcoin.org/pull/2788 But what we can improve here are the following:

### a) Tor Integration to NBitcoin

To communicate with peers over Tor. One way to do this would be to implement Tor communication to NBitcoin, but maybe we should take NBitcoin's code, implement Tor, wait half a year or so until it stabilizes, and then contribute it back to NBitcoin. @lontivero @NicolasDorier opinions?

### b) Bitcoin Core Integration with Tor By Default

If we add Bitcoin Core, then we should not at any circumstances broadcast transactions on the clearnet through it, luckily Tor comes with Wasabi, so connecting the dots here should not be difficult.

## RBF and CPFP

Currently: `checkpassfeecontroldynamic`, next level to target: `checkgoodfeecontrolfull`.
This requires RBF and CPFP. They are problematic from a privacy/wallet fingerprinting point of view, (mainly the former) so if we decide to implement this we must be very careful and a comprehensive elaborate strategy is needed for them.

## Audit

Wasabi did not yet receive an audit. @kristovatlas has expressed interest in doing one some time in the future. When we'll have the resources in the future we can hire him to do so.

# Alternatives

While this can be a direction we could take Wasabi's development, there are others we can consider. Feel free to suggest new ones.  

- Hardware Wallet Integration
- UI And UX Improvement Sprint
- Mixing Improvements: Reverse CoinJoins and P2EP with Reverse CoinJoins
- Accessibility Improvement Sprint: OSX signing, EV HTTPS, rpm, pacman, etc packages and package managers, sweep private key, sweep receive to wrapped segwit, implement bech32 to other wallets/libraries like BitcoinJ, etc...
- Just keep fixing issues in an unorganized way.
- Keep it safe and fix issues tagged with "solution is non-controversial"
- Tacking all the `debug` issues.
- Concentrate on to bring more users on board, which brings us more resources, so we can pay more developers so we can more effectively address issues. For this note that about 7+-2 developers should be the maximum number of devs working on Wasabi, it looks like too much devs make things get out of hand, since over 7 people the administration and collaboration makes the work less and less effective. Refer to: [The Mythical Man-Month](https://en.wikipedia.org/wiki/The_Mythical_Man-Month) and [The Magical Number Seven, Plus or Minus Two](https://en.wikipedia.org/wiki/The_Magical_Number_Seven,_Plus_or_Minus_Two)

**Comment by nopara73** [comment]
@lontivero

# Hardware Wallets

I tried to come up with a hardware wallet integration strategy, but I failed. I don't think we can do that. https://github.com/zkSNACKs/Meta/issues/42

# Reverse CoinJoins

I don't think it should be our next move, I want the code to mature a bit more before we start this. We already deviated way too much from the ZeroLink specification. Let's be more wishful about it.

# Bech32 Compatibility

I'd be fine working on that. We can create a temporary wrapped segwit address to facilitate users to send-sweep money to the wallet and also we could work on implementing bech32 sendability to BitcoinJ, which could probably make the largest impact on the ecosystem: https://github.com/zkSNACKs/WalletWasabi/issues/951

**Comment by crwatkins** [comment]
Although this conversation seems to have moved on, I'll comment on the scoring that I promised yesterday:

> Currently: `checkfailvalidationcentralized`. I believe the above improvement would make it `checkneutralvalidationvariable`. Or is it `checkgoodvalidationfullnode`? Not sure. @crwatkins can you help me out with this?

`checkneutralvalidationvariable` is for hardware wallets that don't do their own validation, so it would most likely be `checkgoodvalidationfullnode` if you provide adequate warnings to users while not in full node validation mode.

>Current: `checkfailenvironmentdesktop`. This is another thing I'm not sure about if we can improve upon it. There are two levels we can target: `checkgoodenvironmenthardware` and `checkpassenvironmenttwofactor`.

`checkgoodenvironmenthardware` is for hardware wallets.  `checkpassenvironmenttwofactor` is for multisig wallets which use a third party to cosign the transaction based on 2FA.

And one you didn't ask about:

>Currently: `checkpassfeecontroldynamic`, next level to target: `checkgoodfeecontrolfull`.
This requires RBF and CPFP.

`checkgoodfeecontrolfull` actually only requires RBF _or_ CPFP support (not both).

Those are the bitcoin.org clarifications as I understand them.

**Comment by nopara73** [comment]
Superseded by https://github.com/zkSNACKs/Meta/issues/68
